src/queue.rs

```diff
@@ -1322,37 +1308,43 @@ pub(crate) mod tests {
     #[test]
     fn test_needs_notification() {
         let m = &GuestMemoryMmap::from_ranges(&[(GuestAddress(0), 0x10000)]).unwrap();
-        let vq = VirtQueue::new(GuestAddress(0), m, 16);
+        let qsize = 16;
+        let vq = VirtQueue::new(GuestAddress(0), m, qsize);
         let mut q = vq.create_queue(&m);
         let avail_addr = vq.avail_start();
 
         // It should always return true when EVENT_IDX isn't enabled.
-        assert_eq!(q.needs_notification(Wrapping(1)), true);
-        assert_eq!(q.needs_notification(Wrapping(2)), true);
-        assert_eq!(q.needs_notification(Wrapping(3)), true);
-        assert_eq!(q.needs_notification(Wrapping(4)), true);
-        assert_eq!(q.needs_notification(Wrapping(5)), true);
+        for i in 0..qsize {
+            q.next_used = Wrapping(i);
+            assert_eq!(q.needs_notification().unwrap(), true);
+        }
 
-        m.write_obj::<u16>(4, avail_addr.unchecked_add(4 + 16 * 2))
+        m.write_obj::<u16>(4, avail_addr.unchecked_add(4 + qsize as u64 * 2))
             .unwrap();
         q.set_event_idx(true);
-        assert_eq!(q.needs_notification(Wrapping(1)), true);
-        assert_eq!(q.needs_notification(Wrapping(2)), false);
-        assert_eq!(q.needs_notification(Wrapping(3)), false);
-        assert_eq!(q.needs_notification(Wrapping(4)), false);
-        assert_eq!(q.needs_notification(Wrapping(5)), true);
-        assert_eq!(q.needs_notification(Wrapping(6)), false);
-        assert_eq!(q.needs_notification(Wrapping(7)), false);
+
+        for i in 0..12 {
+            q.next_used = Wrapping(i);
+            let expected = i == 5 || q.signalled_used.is_none();
+            assert_eq!(q.needs_notification().unwrap(), expected);
+        }
```

Comment:
The spec states:
_For example, if used_event is 0, then a device using VIRTIO_F_EVENT_IDX would send a used buffer notification to the driver after the first buffer is used (and again after the 65536th buffer, etc)._

So how about adding a case for "again after the 65536th buffer"?

Reply:
added